### PR TITLE
Fix/darshan

### PR DESCRIPTION
**Pull request by mtitov**
Updated Darshan integration based on the latest experience of using it

### CODE REVIEW

**Comment by mtitov** [comment]
@andre-merzky ready for review

File: src/radical/entk/tools/darshan.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+# ------------------------------------------------------------------------------
+#
+def enable_darshan(pst: Union[Pipeline, Stage, Task, List[Pipeline]],
+                   darshan_runtime_root: Optional[str] = None,
+                   modules: Optional[List[str]] = None,
+                   env: Optional[Dict[str, str]] = None
+                   ) -> Union[Pipeline, Stage, Task]:
+
+    if not pst:
+        raise ValueError('PST object is not provided')
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:74]: `pst` is not an optional parameter, so this error should never happen?

File: src/radical/entk/tools/darshan.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            return
+
+        darshan_log_dir = DARSHAN_LOG_DIR % {'sandbox': '${RP_TASK_SANDBOX}'}
+        darshan_enable  = (f'LD_PRELOAD="{_darshan_runtime_root}'
+                           '/lib/libdarshan.so" ')
+
+        if src_task.cpu_reqs.cpu_processes == 1:
+            darshan_enable += 'DARSHAN_ENABLE_NONMPI=1 '
+
+        src_task.executable  = darshan_enable + src_task.executable
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:98]: The `executable` should not be a shell command line IMHO.  I would suggest to morph
```py
{  
    'executable': 'foo', 
    'arguments': ['-b', 'bar']
}
```
into
```py
{  
    'executable': '/bin/sh', 
    'arguments': ['-c', 'LD_PRELOAD=/tmp/lib/libdarshan.so foo -b bar']
}
```

File: src/radical/entk/tools/darshan.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        src_task.pre_launch += [f'mkdir -p {darshan_log_dir}']
+        src_task.pre_exec.extend(
+            _darshan_activation_cmds +
+            [f'export DARSHAN_LOG_DIR_PATH={darshan_log_dir}'])
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:102]: This could (should?) got to the task environment?

File: src/radical/entk/tools/darshan.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    return pst
+
+
+# ------------------------------------------------------------------------------
+#
+def get_parsed_data(log: str, target_counters: Union[str, List[str]]) -> set:
+
+    data = set()
+
+    grep_patterns = '-e ' + ' -e '.join(ru.as_list(target_counters))
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:131]: Are we sure that `target_counters` is never an empty list?

File: src/radical/entk/tools/darshan.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    parser_cmd    = (f'darshan-parser {log} | grep {grep_patterns} | '
+                     "awk '{print $5\":\"$6}'")
+    out, err, ret = ru.sh_callout(parser_cmd, env=_darshan_env, shell=True)
+    if ret:
+        print(f'[ERROR] Darshan not able to parse "{log}": {err}')
+    else:
+        for o in out.split('\n'):
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:138]: Nitpick, but I would suggest `for line in ...`

File: src/radical/entk/tools/darshan.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            if not o:
+                continue
+            value, file = o.split(':')
```
Comment (andre-merzky) [inline src/radical/entk/tools/darshan.py:141]: `o.split(':', 1)`

**Review by andre-merzky — APPROVED** [review]
Thanks @mtitov, the code looks reasonable to me.

What I am missing is some explanation what Darshan actually is and what it is used for.  For example: some methods are public methods and will show up in RTD, but don't have a docstring, so the user is left guessing what this is about.

File: src/radical/entk/tools/darshan.py (hunk 1 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py:74]: this is just to catch empty (e.g., `[]`, `{}`, `''`) or `None` values

File: src/radical/entk/tools/darshan.py (hunk 3 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py:102]: yes, thus each task would have its own path to logs

File: src/radical/entk/tools/darshan.py (hunk 4 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py:134]: updated accordingly

File: src/radical/entk/tools/darshan.py (hunk 5 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py]: updated

File: src/radical/entk/tools/darshan.py (hunk 6 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py]: updated

File: src/radical/entk/tools/darshan.py (hunk 2 above)
Comment (mtitov) [reply src/radical/entk/tools/darshan.py:98]: would it be ok to have it within other PR? (I'll convert this comment into task and then will response with a corresponding PR)

p.s. I need to work on the next approach using Darshan config file, and some of the settings might go into it, and I'll keep that in sync with this

Comment (andre-merzky) [reply src/radical/entk/tools/darshan.py:98]: Sure, that works!

**Pull request merged**